src/group/stabchain/builder/random.rs

```diff
@@ -0,0 +1,435 @@
+    //The chain is zero indexed, but this field is 1 indexed.
+    //This is due to the end condition being when the chain is up to date below the index of the first record position, and this would be -1 with zero indexing.
+    //For zero indexing this would have to be a signed type, which doesn't really seem worth it just to require one negative value at the end condition.
+    up_to_date: usize,
+}
+
+impl<P, S, A> StabchainBuilderRandom<P, S, A>
+where
+    P: Permutation,
+    S: MovedPointSelector<P, A::OrbitT>,
+    A: Action<P, OrbitT = usize>,
+{
+    pub fn new(selector: S, action: A) -> Self {
+        StabchainBuilderRandom {
+            current_pos: 0,
+            chain: Vec::new(),
+            selector,
+            action,
+            n: 0,
+            base: Vec::new(),
+            rng: thread_rng(),
+            up_to_date: 1,
+        }
+    }
+
+    fn current_chain(
+        &self,
+    ) -> impl Iterator<Item = &StabchainRecord<P, FactoredTransversalResolver<A>, A>> {
+        self.chain.iter().skip(self.current_pos)
+    }
+
+    pub(super) fn build(self) -> Stabchain<P, FactoredTransversalResolver<A>, A> {
+        Stabchain { chain: self.chain }
+    }
+
+    fn construct_strong_generating_set(&mut self, group: &Group<P>) {
+        //Edge case for trivial group.
+        if group.generators().is_empty() {
+            return;
+        }
+        //Find the largest moved point of any generator, i.e find which size of the symmetric group the generators form a subgroup of.
+        self.n = group
+            .generators
+            .iter()
+            .map(|gen| gen.lmp().expect("Should not be the identity."))
+            .max()
+            .unwrap_or(0);
+        let moved_point = self.selector.moved_point(&group.generators()[0]);
+        //Create the top level record for this chain, and add it to the chain.
+        //TODO check if you should add generators 1 by 1, in case there are redundant generators.
+        let initial_record = StabchainRecord::new(
+            moved_point,
+            group.clone(),
+            factored_transversal_complete_opt(&group, moved_point, &self.action),
+        );
+        self.base.push(moved_point);
+        self.chain.push(initial_record);
+        self.sgc();
+    }
+
+    /// Generate a permutation that is with high probably a schrier generator for the current subgroup.
+    fn random_schrier_generators_as_word(
+        &self,
+        subproducts: usize,
+        coset_representatives: usize,
+        gens: &[P],
+    ) -> Vec<Vec<P>> {
+        //Sum of all "depths". In reality the transversal doesn't have a depth, so we use this as a upper bound.
+        let t = self
+            .current_chain()
+            .map(|record| (record.transversal.len() as f64).log2())
+            .sum::<f64>()
+            .floor() as usize;
+        let record = &self.chain[self.current_pos];
+        let k = rand::Rng::gen_range(&mut self.rng.clone(), 0, gens.len() / 2 + 1);
```

Comment:
Can you make it 1 + gens.len() / 2 ? 
Sorry it is just a tiny bit clearer to me

Reply:
Done